### ISSUE TITLE
[building a potential Edge AI solution] As an ecologist, what are you looking for to extract from camera traps?

### ISSUE THREAD

**Issue by ejri** [microsoft/CameraTraps#176]
Hey all, 

Not sure if it falls under an issue (I'm happy to post or x-post it somewhere more suitable). This repo may be the best place to ask some questions about what's needed to be integrated in camera traps, since individuals with technical backgrounds, with some understanding on research needs, are likely to stumble upon the repo.    

I'm an AI for Earth grantee (agritech and their integration within smart cities). 

I learned about camera traps during the AI for Earth Summit, and coincidentally, as a part of an online competition (Microsoft's Discover AI Upskilling Journey) where organizers have asked participants to look into helping out with camera traps. 

However, what is not particularly clear is, what exactly are ecologists looking to extract out of camera trap images? 
Animals (detection and classifying which animal)?
Other objects related to poaching (people, cars, trucks)? 
Other features (weather conditions, time stamps of when detection happens)? 

Understanding what needs to be extracted can help us (researchers and hobbyists) understand what needs to be developed, as well as tailor the hardware and integration to better fit what is needed.  

### Overview of my project: 

The goal of this project is to highlight how effective Edge AI can be in eliminating/filtering out a significant number of inter-steps between data collection and obtaining classification and detection results, potentially along with other data/features that need to be collected or extracted. 

![Presentation1](https://user-images.githubusercontent.com/2915475/85766317-c81a3b00-b6cb-11ea-804a-97fb11b14fc7.jpg)
Picture legend:
green arrows -> working
red arrows -> not working as intended
blue arrows -> have not tried it yet int this project, but tested out and working in other projects

**More information:**
The project aims to tie-in features from machine learning and internet of things with hardware devices, similar in some extent to what is currently used on the ground, in order to reduce the overall time spent ecologists spend on classifications, collecting data, etc... As someone who has never interacted with ecologists, it's difficult to understand what is currently needed for them to better allocate their time and focus on research aspects, rather than countless hours on data labelling. 

**Model**: I trained a custom object detection and classification (Yolov4) model for 17 classes, which are all the classes of animals that I was able to manually make out in the [MegaDetector demo video](dolphinvm.westus2.cloudapp.azure.com/video/detector_video.html). The classes are Armadillo, Bear, Bird, Bull, [Car], Cat, Cattle, Deer, Dog, Fox, Monkey, [Person], Pig, Raccoon, Sheep, Tiger, [Truck]. I added a few classes in [brackets], just in case someone would find the model to use as is.

The images used for training are obtained from [Open Images Dataset V6](https://storage.googleapis.com/openimages/web/index.html)

Then, I used model to run object detection and classification on some test images (the sheep image in the overview is one of the test images used) and the MegaDetector demo video. 

Testing images (after ~5hrs training):
Yolov4 TFLite (converted from yolov4)
<img width="458" alt="yolov4 TFLite" src="https://user-images.githubusercontent.com/2915475/85774663-e97f2500-b6d3-11ea-825a-80db3fcfe5dd.png">

Yolov4
<img width="456" alt="yolov4" src="https://user-images.githubusercontent.com/2915475/85774681-edab4280-b6d3-11ea-94be-f7f72cc2fefb.png">

Video:
Running the yolov4 model on the MegaDetector demo video (less than 2 minutes for the model to detect and classify the video).

It seems to be pretty ok for some parts, needs more training for other parts.
Some examples for detection/classification:
<img width="721" alt="Screen Shot 2020-06-25 at 10 56 42 AM" src="https://user-images.githubusercontent.com/2915475/85775921-14b64400-b6d5-11ea-8c3e-315f462c81d9.png">
<img width="722" alt="Screen Shot 2020-06-25 at 10 56 16 AM" src="https://user-images.githubusercontent.com/2915475/85775944-1a138e80-b6d5-11ea-849d-a326dc5bdaae.png">
<img width="722" alt="Screen Shot 2020-06-25 at 10 55 52 AM" src="https://user-images.githubusercontent.com/2915475/85775948-1b44bb80-b6d5-11ea-99ea-d3016c64083c.png">
<img width="722" alt="Screen Shot 2020-06-25 at 10 58 01 AM" src="https://user-images.githubusercontent.com/2915475/85776147-45967900-b6d5-11ea-8bfc-6ef6726be858.png">

Some examples for bad detection/classification (model needs more training?):
<img width="721" alt="Screen Shot 2020-06-25 at 10 59 07 AM" src="https://user-images.githubusercontent.com/2915475/85776256-6068ed80-b6d5-11ea-9ae8-58e442b12360.png">
<img width="721" alt="Screen Shot 2020-06-25 at 10 57 10 AM" src="https://user-images.githubusercontent.com/2915475/85776271-652da180-b6d5-11ea-80de-dc921dddfe96.png">

**Extracting more information from images**

I was reading that ecologists also look at camera trap images to extract more information about the ecosystem, such as weather conditions. 

My hand-waving approach was to use Custom Vision to train weather data, and I stumbled upon a [Multi-class Weather Dataset for Image Classification.] (https://data.mendeley.com/datasets/4drtyfjtfy/1)

<img width="1096" alt="Screen Shot 2020-06-25 at 11 18 21 AM" src="https://user-images.githubusercontent.com/2915475/85777580-9a86bf00-b6d6-11ea-8cc9-52e8ac4bed80.png">
<img width="1097" alt="Screen Shot 2020-06-25 at 11 19 00 AM" src="https://user-images.githubusercontent.com/2915475/85777595-9e1a4600-b6d6-11ea-9d60-4058f7222e11.png">

However, testing the model on some images, the model is not the best for camera-traps, and even though some classifications were on point, but the accuracy is really low. 

<img width="1095" alt="Screen Shot 2020-06-25 at 11 19 22 AM" src="https://user-images.githubusercontent.com/2915475/85778016-fea98300-b6d6-11ea-8679-33f0ece71d55.png"> 

<img width="1098" alt="Screen Shot 2020-06-25 at 11 25 14 AM" src="https://user-images.githubusercontent.com/2915475/85777984-f94c3880-b6d6-11ea-8f38-aa1006531ae3.png">

**Another hand-waving approach to extract weather data from camera-trap images (haven't implemented this yet):**

It seems that some images display temperature data in each image, so another approach is to extract temperature information as well as time/date stamps. Then use historical weather datasets (such as openweather API) to grab other weather features such as humidity, wind direction/speed, weather conditions (rainy, cloudy, partially sunny, sunny). 

**Raspberry pi as an Edge AI Device**

<img width="783" alt="Screen Shot 2020-06-22 at 5 38 48 PM" src="https://user-images.githubusercontent.com/2915475/85784418-1edc4080-b6dd-11ea-89a9-8ee66688ceb0.png">

As implied earlier, the main reason to convert the Yolov4 model to a TensorFlow Lite model is for the model to run on the raspberry pi itself. This means that as an image or video is taken, object detection and classification happens on the device itself, where the results can be either uploaded directly to the internet or saved on an SD card. 

As the weather information is relatively more tricky, integrating relatively cheap sensors may provide a more custom overview of the ecosystem. For example, it is possible to add sensors to collect data about emissions such as COx, NOx, NH3, VOCs as well as air quality. It would also allow for more customizability when it comes to choosing a camera that may be a better fit for custom applications, as well as detection sensors (in this example, pyroelectric/ infrared sensor in addition to an ultrasonic sensor).

**Comment by agentmorris** [comment]
Thanks for sharing!  I will try to answer one of your questions here:

> What exactly are ecologists looking to extract out of camera trap images?

It varies from project to project, but some common themes:

- Maybe 85% of our collaborators are focused on estimating wildlife populations.  Among those, most go through every image and quantify the number of individuals present for every species present.  Some add additional features (e.g. gender, behavior, antler size), but species is the least common denominator.  Most of these projects will also quantify humans and vehicles, but as a secondary goal.
- Most of the remaining 15% are primarily focused on humans and vehicles for offline surveys, e.g. park managers who want to quantify trail use.
- A small but growing number of users are interested in detecting humans and vehicles in real-time for anti-poaching applications.  This is a very small fraction of global camera trap use, but is growing fast, and is a critical case for AI.
- Estimating weather from camera traps is something we've encountered, but it's rare.

I *think* these are relatively representative of the larger camera trap community, but take what I say with a grain of salt, there are a number of biases in whom we end up in contact with.

Hope that helps!

-Dan

**Comment by agentmorris** [comment]
Closing because this is a discussion, not an issue; happy to keep discussing here, but preventing this from showing up as an open issue.

**Comment by abfleishman** [comment]
@ejri I think Dan is right in what many ecologists are looking to extract, _currently_ but I think that is because of the amount of work that is needed to extract each piece of information manually.  If there was a system to automatically classify weather I bet a bunch of researchers would want to use it! Often times camera trap surveys are limited by what people think they can handle in terms of data review.  For example, I might choose to put out only 15-20 camera because it costs $1000 per camera to review the images and so my budget can only handle so many cameras, or I might choose to use motion-triggered events as opposed to timelapse settings because timelapse (1 frame every 1-2 second) generates WAY too many images to review manually. But if you can use AI to process the images in a cost-effective way, then I might choose to scale my survey 10x or 100x! 

Another way that I have used camera traps to measure breeding success in birds.  A camera is focused on a nest and takes timelapse photos and you can extract info such as the number of adults, eggs, or chicks, date of laying, date of hatching, date of egg or chick loss/death, and/or fledging.  And from these data you can extract info about nest predators and conditions that might have lead to the loss of a chick or egg.  This is a great example of the kind of project that limits the number of cameras deployed because it is too much work to manually review the data, but often the sample size (in this case number of nests monitored) is often much smaller than a statistician would recommend to make inferences across your population.

**Comment by abfleishman** [comment]
Another thing that would be useful to detect in images is image quality e.g. is it blurry, the lens fogged up, completely black or white (under/overexposed).  This would be useful to remove images that are not useful for analysis (help quantify monitoring effort ) and also in a real-time context, alert the user that your camera needs attention (maybe a field visit)

**Comment by beerys** [comment]
Other things that I have heard requests for over the years:

1. Counts for each species
2. Classification of behavior
2. Estimate of animal distance from the camera
3. Estimate of animal age, or even just young/old classification
4. Detection of disease in animals (through fur loss or skin conditions), to help track disease spreads
5. Detection of molting
6. Habitat categorization (ie classifying the plants that appear in a camera location, or tracking their change over time)
7. Detection of invasive plant species

Just a few that were at the top of my head!i

**Comment by ejri** [comment]
Thank you for your answers and insight, it certainly helps!

I'd imagine one would need a high resolution camera for this:
> 4. Detection of disease in animals (through fur loss or skin conditions), to help track disease spreads
> 5. Detection of molting
> 6. Habitat categorization (ie classifying the plants that appear in a camera location, or tracking their change over time)
> 7. Detection of invasive plant species

We're kind of looking at 6 and 7 as a part of our project (the agriculture one) as well, in addition to detecting diseases on plants, to try and check to what level pesticides are working, and potentially reduce pesticides usages by hopefully a huge margin (along with associated emissions).


![Picture1](https://user-images.githubusercontent.com/2915475/85907713-3ee82e80-b7c7-11ea-88d0-6bcecf8f88cd.png)
